oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/profile/analyze/ProfileStackNode.java

```diff
@@ -0,0 +1,242 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.profile.analyze;
+
+import com.google.common.base.Objects;
+import org.apache.skywalking.oap.server.core.query.entity.ProfileStackElement;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Work for profiling stacks, intermediate state of the {@link ProfileStackElement} and {@link ProfileStack}
+ */
+public class ProfileStackNode {
+
+    // stack code signature
+    private String codeSignature;
+    // owner list
+    private List<ProfileStack> owners;
+    // children nodes
+    private List<ProfileStackNode> children;
+    // include the execution time of children
+    private int duration;
+
+    /**
+     * create new empty, un-init node
+     * @return
+     */
+    public static ProfileStackNode createEmptyNode() {
+        ProfileStackNode emptyNode = new ProfileStackNode();
+        emptyNode.owners = new LinkedList<>();
+        emptyNode.children = new LinkedList<>();
+        return emptyNode;
+    }
+
+    /**
+     * accumulate {@link ProfileStack} to this tree, it will invoke on the tree root node
+     * @param stack
+     */
+    public void accumulateFrom(ProfileStack stack) {
+        List<String> stackList = stack.getStack();
+        if (codeSignature == null) {
+            codeSignature = stackList.get(0);
+        }
+        // increase self
+        this.increase(stack);
+
+        // handle stack children
+        ProfileStackNode parent = this;
+        for (int childDepth = 1; childDepth < stackList.size(); childDepth++) {
+            String elementCodeSign = stackList.get(childDepth);
+
+            // find same code signature children
+            ProfileStackNode childElement = null;
+            for (ProfileStackNode child : parent.children) {
+                if (Objects.equal(child.codeSignature, elementCodeSign)) {
+                    childElement = child;
+                    break;
+                }
+            }
+
+            if (childElement != null) {
+                // increase child node
+                childElement.increase(stack);
+                parent = childElement;
+            } else {
+                // add children
+                ProfileStackNode childNode = createEmptyNode();
+                childNode.codeSignature = elementCodeSign;
+                childNode.increase(stack);
+
+                parent.children.add(childNode);
+                parent = childNode;
+            }
+        }
+    }
+
+    /**
+     * combine from other {@link ProfileStackNode}
+     * @param node
+     * @return
+     */
+    public ProfileStackNode combine(ProfileStackNode node) {
+        // combine this node
+        this.combineThisNode(node);
+
+        // merge tree using pre order
+        LinkedList<Pair<ProfileStackNode, List<ProfileStackNode>>> stack = new LinkedList<>();
```

Comment:
- Q1. Which algorithm are you using for tree merging?
- Q2. Are you using the stack to avoid `recursion`? The codes seem understandable, but not very clear.

Reply:
Q1: Using LDR to traversal tree node, and merge children nodes to the parent node.
Q2: Yes. If the stack has large tree stack depth, It will throw StackOverflowError.